File: src/update_wt.f
C update weights
C output: weights_update
      subroutine update_wt(n, weights, y, f, cfun, dfun, s, delta,
     +     weights_update)
       implicit none
       integer cfun, dfun, i, n
       double precision weights(n),weights_update(n), y(n),f(n),
     + u(n),z(n), v(n), s, delta
       external compute_v, compute_z, compute_u 
 
C     compute u
             call compute_u(dfun, n, y, f, u)
C     compute z=s(u), s is not used for the selected dfun
             call compute_z(dfun, n, u, z, s)
C     compute derivative of -g(z)
             call compute_v(cfun, n, z, s, delta, v)
C     update weights
             do 10 i=1, n
              weights_update(i)=-weights(i)*v(i)
10         continue
      return
      end

File: R/update_wt.R
### this is different from src/update_wt in that the inputs are different 
update_wt <- function(y, ypre, weights, cfun, s, dfun, delta=0.0001){
    if(dfun %in% 4:7 && !setequal(unique(y), c(-1, 1))) stop("y must be -1/1 for binary classification")
    n <- length(y)
    if(missing(weights)) weights <- rep(1, n)
    if(cfun==6)
        if(s > 1) delta <- (s-1)/2 else
                                       if(s==1) delta <- 0 else{
                                                               if(is.null(delta)) stop("delta must be provided")
                                                               if(delta <= 0) stop("delta must be positive")
                                                           }
    .Fortran("update_wt",
             n=as.integer(n),
             weights=as.double(weights),
             y=as.double(y),
             f=as.double(ypre),
             cfun=as.integer(cfun),
             dfun=as.integer(dfun),
             s=as.double(s),
             delta=as.double(delta),
             weights_update=as.double(rep(0, n)),
             PACKAGE="mpath")$weights_update
}
